File: find_bluefs_index.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
序列连续性检测工具 v1.1

功能说明：
    检测fix_search_map中的seq连续性，并在header_map中查找可用版本
    支持百万级数据的高效处理

作者：AI助手
创建日期：2025-04-26
版本：1.1.0
"""

import sys
from collections import OrderedDict
from typing import Dict, List, Optional, Tuple

def read_seq_file(file_path: str) -> Dict[int, int]:
    """读取seq文件并构建字典
    
    Args:
        file_path: 文件路径
        
    Returns:
        Dict[int, int]: {seq: disk_offset}字典
        
    示例：
        >>> read_seq_file("fix_search_map")
        {1: 100, 2: 200, 3: 300}
    """
    seq_dict = OrderedDict()
    try:
        with open(file_path, 'r') as f:
            for line in f:
                stripped = line.strip()
                if not stripped:
                    continue
                parts = stripped.split()
                if len(parts) != 4 or parts[0] != 'Found':
                    continue
                try:
                    seq = int(parts[2].split(':')[1])
                    offset = int(parts[3].split(':')[1])
                    seq_dict[seq] = offset
                except (IndexError, ValueError):
                    continue
    except FileNotFoundError:
        print(f"Error: File not found - {file_path}", file=sys.stderr)
        sys.exit(1)
    return seq_dict

def find_continuous_sequence(fix_dict: Dict[int, int], header_dict: Dict[int, int]) -> Optional[Tuple[int, List[int]]]:
    """查找连续序列并返回可用版本
    
    Args:
        fix_dict: fix_search_map字典
        header_dict: header_map字典
        
    Returns:
        Optional[Tuple[int, List[int]]]: (起始seq, disk_offsets列表)或None
    """
    if not fix_dict:
        print("Error: fix_search_map is empty", file=sys.stderr)
        return None
    
    max_seq = max(fix_dict.keys())
    sorted_seqs = sorted(fix_dict.keys(), reverse=True)
    
    break_point = None
    for i in range(1, len(sorted_seqs)):
        if sorted_seqs[i-1] - sorted_seqs[i] > 1:
            break_point = sorted_seqs[i]
            break
    
    if break_point is None:
        # 全部连续的情况
        return (max_seq, [fix_dict[max_seq]])
    
    candidate_seq = break_point + 1
    if candidate_seq in header_dict:
        # 找到可用版本
        start_seq = candidate_seq
        end_seq = max_seq
        return (start_seq, [fix_dict[s] for s in range(start_seq, end_seq+1)])
    
    # 未找到可用版本，调整max_seq
    new_max = break_point
    if new_max in fix_dict:
        return find_continuous_sequence({k:v for k,v in fix_dict.items() if k >= new_max}, header_dict)
    return None

def main():
    if len(sys.argv) != 3:
        print("Usage: python3 seq_checker.py fix_search_map header_map", file=sys.stderr)
        sys.exit(1)
    
    fix_dict = read_seq_file(sys.argv[1])
    header_dict = read_seq_file(sys.argv[2])
    
    result = find_continuous_sequence(fix_dict, header_dict)
    
    if result:
        start_seq, offsets = result
        print(f"Found continuous sequence starting at seq {start_seq}:")
        for offset in offsets:
            print(offset)
    else:
        print("No valid sequence found", file=sys.stderr)

if __name__ == "__main__":
    main()

File: find_bluefs_index.cpp
#include <stdint.h>

#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

using namespace std;

// 读取序列文件并构建有序字典
std::vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>> read_seq_file(const std::string& file_path) {
    std::vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>> seq_list;
    ifstream file(file_path);

    if (!file.is_open()) {
        throw runtime_error("Error: File not found - " + file_path);
    }

    string line;
    while (getline(file, line)) {
        istringstream iss(line);
        string word;
        vector<string> parts;

        // 分割单词
        while (iss >> word) {
            parts.push_back(word);
        }

        if (parts.empty() || parts.size() != 2) {
            continue;
        }

        // 第一个是disk_offset
        uint64_t disk_offset = std::stoull(parts[0].c_str());

        // 第二个是seq
        uint64_t seq = std::stoull(parts[1].c_str());

        seq_list.push_back({disk_offset, seq});
    }

    // 按照disk_offset排序
    std::sort(seq_list.begin(), seq_list.end(), [](const auto& a, const auto& b) { return a.second < b.second; });

    return seq_list;
}

// 查找连续序列并返回可用版本
pair<uint64_t /*disk_offset_header*/, vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>> /*next_seq*/> find_continuous_sequence(
    const std::vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>>& fix_list, const std::vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>>& header_list) {
    if (fix_list.empty()) {
        throw runtime_error("Error: fix_search_map is empty");
    }

    pair<uint64_t /*disk_offset_header*/, vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>> /*next_seq*/> ans;

    // 获取最大序列号
    uint64_t max_seq = fix_list.back().second;

    std::cout << "max_seq: " << max_seq << std::endl;

    ans.first = UINT64_MAX;

    // 查找断点
    bool has_find = false;
    uint64_t break_point = max_seq + 1;
    for (size_t i = fix_list.size() - 1; i > 0; i--) {
        // std::cout << "check seq: " << fix_list[i].second << "max_seq: " << max_seq << " break: " << break_point << std::endl;
        if (fix_list[i].second != break_point - 1) {
            break_point = break_point;  // point to self. it's jump_seq
            has_find = true;
            break;
        }
        break_point = fix_list[i].second;
    }

    // 处理完全连续的情况
    if (!has_find) {
        std::cout << "not found break point" << std::endl;
        return ans;
    } else {
        std::cout << "break_point: " << break_point << std::endl;
    }

    // 检查候选序列
    for (auto& p : header_list) {
        if (p.second == break_point) {
            ans.first = p.first;
            std::cout << "find header jump_seq: " << p.second << std::endl;
            std::cout << "find header disk_offset: " << p.first << std::endl;
            break;
        }
    }

    for (size_t i = fix_list.size() - 1; i > 0; i--) {
        if (fix_list[i].second >= break_point) {
            ans.second.push_back(fix_list[i]);
        }
    }

    // 对ans.second按照seq来排序
    std::sort(ans.second.begin(), ans.second.end(), [](const auto& a, const auto& b) { return a.second < b.second; });

    return ans;
}

void write_offset_seq(const std::string& file_path, const std::vector<std::pair<uint64_t /*disk_offset*/, uint64_t /*seq*/>>& seq_list) {
    ofstream file(file_path);
    if (!file.is_open()) {
        throw runtime_error("Error: Unable to open file for writing - " + file_path);
    }

    for (const auto& p : seq_list) {
        file << p.first << " " << p.second << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " fix_search_map header_map" << endl;
        return 1;
    }

    try {
        // 读取输入文件
        auto fix_dict = read_seq_file(argv[1]);
        write_offset_seq("fix_search_osd_4_map_sort", fix_dict);
        auto header_dict = read_seq_file(argv[2]);

        // 输出调试信息
        cout << "fix_dict size: " << fix_dict.size() << endl;

        // 查找连续序列
        auto ans = find_continuous_sequence(fix_dict, header_dict);

        // 输出结果
        std::cout << "first jump offet: " << ans.first << std::endl;
        std::cout << "next seq: " << std::endl;
        for (const auto& p : ans.second) {
            std::cout << p.first << " " << p.second << std::endl;
        }
    } catch (const exception& e) {
        cerr << e.what() << endl;
        return 1;
    }

    return 0;
}